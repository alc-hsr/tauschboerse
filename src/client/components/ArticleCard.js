import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import { Card, CardMedia, CardTitle, CardText, CardActions } from 'material-ui/Card';

import EditIcon from 'material-ui/svg-icons/editor/mode-edit';
import AccountIcon from 'material-ui/svg-icons/action/account-circle';

import ArticleStatusComponent from './ArticleStatusTag';
import AvatarTag from './AvatarTag';
import PhotoPlaceholder from './PhotoPlaceholder';

import './ArticleCard.css';

export default class ArticleCard extends React.Component {

    static propTypes = {
        article: PropTypes.object.isRequired,
        actions: PropTypes.array.isRequired
    };

    static defaultProps = {
        actions: []
    };

    render() {
        const { article, actions } = this.props;
        const { title, description, status, owner, categories, photos, created } = article;
        const name = (owner) ? owner.name : '';
        const categoryNames = (categories) ? categories.map(category => category.name).join(', ') : '';
        const photoSource = (photos && photos.length > 0) ? photos[0].url : null;
        return (
            <div className="article-card">
                <Card>
                    <div className="article-card__header">
                        <AvatarTag text={name} icon={<AccountIcon/>}/>
                        <AvatarTag text={`${moment(created).format('DD.MM.YYYY | HH:mm')}`} icon={<EditIcon/>}/>
                        <ArticleStatusComponent status={status}/>
                    </div>
                    <CardMedia overlay={<CardTitle title={title} subtitle={categoryNames}/>}>
                        <div className="article-card__image-container">
                            {photoSource ? <img className="article-card__image" src={photoSource} alt={title}/> : <PhotoPlaceholder width={300} height={300}/>}
                        </div>
                    </CardMedia>
                    <CardText className="article-card__text">{description}</CardText>
                    <CardActions>{actions}</CardActions>
                </Card>
            </div>
        );
    }
}
