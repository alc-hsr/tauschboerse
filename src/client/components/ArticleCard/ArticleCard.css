.article-card {
    margin: 10px;
    max-width: 450px;
    min-width: 250px;
}

.article-card:hover {
    box-shadow: 0 6px 20px 0 gray;
    transition: all 0.5s linear;
}

.article-card__header {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: flex-start;
    flex-wrap: wrap;
    align-content: flex-start;
}

.article-card__image-container {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 377px;
    max-width: 450px;
}

.article-card__image {
    max-width: 100%;
    max-height: 100%;
}

.article-card__text {
    white-space: pre-wrap;
    height: 50px;
    overflow: hidden;
}
