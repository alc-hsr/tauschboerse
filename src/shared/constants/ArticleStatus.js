const STATUS_FREE = 'FREE';
const STATUS_DEALING = 'DEALING';
const STATUS_DEALED = 'DEALED';
const STATUS_DELETED = 'DELETED';

function getShortDescription(theStatus) {
    switch (theStatus) {
        case STATUS_FREE:
            return 'Frei';
        case STATUS_DEALING:
            return 'In Verhandlung';
        case STATUS_DEALED:
            return 'Getauscht';
        case STATUS_DELETED:
            return 'Gelöscht';
        default:
            return theStatus;
    }
}

function getDescription(theStatus) {
    switch (theStatus) {
        case STATUS_FREE:
            return 'Artikel steht frei zur Verfügung';
        case STATUS_DEALING:
            return 'Der Artikel wird gehandelt';
        case STATUS_DEALED:
            return 'Der Artikel wurde getauscht';
        case STATUS_DELETED:
            return 'Der Artikel wurde gelöscht';
        default:
            return theStatus;
    }
}

function getColor(theStatus) {
    switch (theStatus) {
        case STATUS_FREE:
            return '#4444FF';
        case STATUS_DEALING:
            return '#FFA000';
        case STATUS_DEALED:
            return '#00BCD4';
        case STATUS_DELETED:
            return '#DD2C00';
        default:
            return '#000000';
    }
}

const validArticleStatuses = [STATUS_FREE, STATUS_DEALING, STATUS_DEALED, STATUS_DELETED];

function isValidArticleStatus(val) {
    return validArticleStatuses.indexOf(val) >= 0;
}

module.exports = {
    STATUS_FREE,
    STATUS_DEALING,
    STATUS_DEALED,
    STATUS_DELETED,
    getShortDescription,
    getDescription,
    getColor,
    isValidArticleStatus
};
