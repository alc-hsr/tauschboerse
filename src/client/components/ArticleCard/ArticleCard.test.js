import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import toJson from 'enzyme-to-json';

import getMuiTheme from 'material-ui/styles/getMuiTheme';

import ArticleCard from './ArticleCard';
import ArticleStatus from '../../../shared/constants/ArticleStatus';

const muiTheme = getMuiTheme();

Enzyme.configure({ adapter: new Adapter() });

const createArticle = () => {
    return {
        _id: '1',
        title: 'Fussballschuhe',
        description: 'Fussballschuhe, kaum gebraucht...',
        created: new Date('2017-12-16'),
        status: ArticleStatus.STATUS_FREE
    };
};

describe('ArticleCard', () => {
    test('Test that the ArticleCard renders correctly (snapshot).', () => {
        const article = createArticle();
        const tree = shallow(<ArticleCard article={article}/>, { context: { muiTheme } });
        expect(toJson(tree)).toMatchSnapshot();
    });
});
