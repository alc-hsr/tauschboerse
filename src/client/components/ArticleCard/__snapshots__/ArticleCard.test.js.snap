// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard Test that the ArticleCard renders correctly (snapshot). 1`] = `
<div
  className="article-card"
>
  <Card
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
  >
    <div
      className="article-card__header"
    >
      <AvatarTag
        icon={<ActionAccountCircle />}
        text=""
      />
      <AvatarTag
        icon={<EditorModeEdit />}
        text="16.12.2017 | 01:00"
      />
      <ArticleStatusTag
        status="FREE"
      />
    </div>
    <CardMedia
      overlay={
        <CardTitle
          subtitle=""
          title="Fussballschuhe"
        />
      }
    >
      <div
        className="article-card__image-container"
      >
        <PhotoPlaceholder
          height={300}
          width={300}
        />
      </div>
    </CardMedia>
    <CardText
      className="article-card__text"
    >
      Fussballschuhe, kaum gebraucht...
    </CardText>
    <CardActions />
  </Card>
</div>
`;
